### PR TITLE
external: auto switch reader mode 

### LINKED ISSUE #45719: Support global sort on S3
## Enhancement
This PR collects some PRs to support global sort. 
TiDB servers can sort a large amount of data on shared storage(S3 for now), then ingect into TiKV without overlap.

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #45719

Problem Summary:
Support switching read mode automatically so the speed does not degrade for the hot-point.

### What is changed and how it works?
Record the file access stat, and switch reader mode if necessary.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/byte_reader.go [hunk 1]
```diff
@@ -178,6 +178,7 @@ func (r *byteReader) reload() error {
 	to := r.useConcurrentReader.Load()
 	now := r.useConcurrentReaderCurrent.Load()
 	if to != now {
+		logutil.BgLogger().Info("switch reader mode", zap.Bool("use concurrent mode", to))
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/byte_reader.go:181]:
```suggestion
		logutil.Logger(r.ctx).Info("switch reader mode", zap.Bool("use concurrent mode", to))
```

File: br/pkg/lightning/backend/external/iter_test.go [hunk 2]
```diff
@@ -337,3 +346,73 @@ func BenchmarkPointerT(b *testing.B) {
...
+
+func testMergeIterSwitchMode(t *testing.T, f func([]byte, int) []byte) {
+	st, clean := NewS3WithBucketAndPrefix(t, "test", "prefix/")
+	defer clean()
+
+	// Prepare
+	writer := NewWriterBuilder().
+		SetPropKeysDistance(100).
+		SetMemorySizeLimit(512*1024).
+		Build(st, "testprefix", strconv.Itoa(0))
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/iter_test.go:384]:
```suggestion
		Build(st, "testprefix", "0"))
```

File: br/pkg/lightning/backend/external/kv_reader.go [hunk 3]
```diff
@@ -42,7 +42,6 @@ func newKVReader(
 	}
 	br, err := newByteReader(ctx, sr, bufSize, store, name, false)
 	if err != nil {
-		br.Close()
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/kv_reader.go:45]: where do we Close now?

File: br/pkg/lightning/backend/external/iter.go [hunk 4]
```diff
@@ -198,6 +202,23 @@ func (i *mergeIter[T, R]) next() bool {
 	var zeroT T
 	i.curr = zeroT
 	if i.lastReaderIdx >= 0 {
+		cnt, ok := i.checkHotPointMap[i.lastReaderIdx]
+		if !ok {
+			i.checkHotPointMap[i.lastReaderIdx] = 1
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/iter.go:207]: we can rely on the zero value when map entry is not found, maybe

```
i.checkHotPointMap[i.lastReaderIdx] = i.checkHotPointMap[i.lastReaderIdx] + 1
```

File: br/pkg/lightning/backend/external/iter.go [hunk 5]
```diff
@@ -198,6 +202,23 @@ func (i *mergeIter[T, R]) next() bool {
...
+		} else {
+			i.checkHotPointMap[i.lastReaderIdx] = cnt + 1
+		}
+		i.checkHotPointCnt++
+
+		if i.checkHotPointCnt == 1000 {
+			i.checkHotPointCnt = 0
+			for idx, cnt := range i.checkHotPointMap {
+				(*i.readers[idx]).setReadMode(cnt > 500)
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/iter.go:216]: seems byteReader is not concurrently used, we don;t need atomic.Bool to represent useConcurrentReader

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/backend/external/iter.go [hunk 6]
```diff
@@ -198,6 +202,18 @@ func (i *mergeIter[T, R]) next() bool {
...
+		i.checkHotPointMap[i.lastReaderIdx] = i.checkHotPointMap[i.lastReaderIdx] + 1
+		i.checkHotPointCnt++
+
+		if i.checkHotPointCnt == 1000 {
+			i.checkHotPointCnt = 0
+			for idx, cnt := range i.checkHotPointMap {
+				(*i.readers[idx]).setReadMode(cnt > 500)
+			}
+			i.checkHotPointCnt = 0
+			i.checkHotPointMap = make(map[int]int)
```
Comment (okJiang) [inline br/pkg/lightning/backend/external/iter.go:214]:
```suggestion
			for idx, cnt := range i.checkHotPointMap {
				(*i.readers[idx]).setReadMode(cnt > 500)
			}
			i.checkHotPointCnt = 0
			i.checkHotPointMap = make(map[int]int)
```

File: br/pkg/lightning/backend/external/iter.go [hunk 7]
```diff
@@ -198,6 +202,18 @@ func (i *mergeIter[T, R]) next() bool {
 	var zeroT T
 	i.curr = zeroT
 	if i.lastReaderIdx >= 0 {
+		i.checkHotPointMap[i.lastReaderIdx] = i.checkHotPointMap[i.lastReaderIdx] + 1
+		i.checkHotPointCnt++
+
+		if i.checkHotPointCnt == 1000 {
```
Comment (okJiang) [inline br/pkg/lightning/backend/external/iter.go:208]: How about expressing 1000 in constants? And add some comments for it.

File: br/pkg/lightning/backend/external/iter.go [hunk 8]
```diff
@@ -198,6 +202,18 @@ func (i *mergeIter[T, R]) next() bool {
...
+			i.checkHotPointCnt = 0
+			for idx, cnt := range i.checkHotPointMap {
+				(*i.readers[idx]).setReadMode(cnt > 500)
```
Comment (okJiang) [inline br/pkg/lightning/backend/external/iter.go:211]: 500 ditto

**Review by lance6716 — APPROVED** [review]

File: br/pkg/lightning/backend/external/iter.go [hunk 9]
```diff
@@ -68,11 +69,13 @@ func (h *mergeHeap[T]) Pop() interface{} {
...
-	readers       []*R
-	curr          T
-	lastReaderIdx int
-	err           error
+	h                mergeHeap[T]
+	readers          []*R
+	curr             T
+	lastReaderIdx    int
+	err              error
+	checkHotPointMap map[int]int
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/iter.go:77]: Maybe rename to `hotspotMap`?

**Review by tangenta — COMMENTED** [review]
Rest LGTM

File: br/pkg/lightning/backend/external/iter.go (hunk 9 above)
Comment (okJiang) [reply br/pkg/lightning/backend/external/iter.go:77]: I didn't understand the meaning of this name at first glance too.

**Review by tangenta — APPROVED** [review]

**Pull request merged**